### DISCUSSION TITLE
`gatsby-source-contentful@4.0.0` Rich Text Changes/Migration

### DISCUSSION THREAD

**Discussion by LekoArts** [gatsbyjs/gatsby#28098] [Umbrella Discussions]
## Description

With [version 4.0.0](https://github.com/gatsbyjs/gatsby/blob/master/packages/gatsby-source-contentful/CHANGELOG.md#400-next0-2020-11-09) the Rich Text experience was changed. Please use this umbrella discussion to share migrations or ask questions or report possible bugs.

@axe312ger can share more information then.

## Related issues

https://github.com/gatsbyjs/gatsby/issues/28052
https://github.com/gatsbyjs/gatsby/issues/28074

**Comment by mikaelkristiansson** [comment]
Not sure if this is a bug or just a miss in the documentation but when setting a reference to an `Asset` it will not include it unless you add the key `__typename` in the graphQL query.
so like this:
```
body {
      raw
      references {
        ... on ContentfulAsset {
          contentful_id
          __typename
          file {
            url
          }
        }
      }
    }
```

this is because if you use `richTextRender` function it will divide the types by that key.
https://github.com/gatsbyjs/gatsby/blob/master/packages/gatsby-source-contentful/src/rich-text.js#L30

It might have to do something with this comment? https://github.com/gatsbyjs/gatsby/pull/25249#issuecomment-665043279
I havent tried out the `useNameForId: false` yet though.

**Reply by axe312ger** [reply]
I'd love to include id & type automatically, lets see if we find a solution :)

**Reply by axe312ger** [reply]
Looks like we can, I'll create an issue out of your comment :)

**Reply by gusliedke** [reply]
If you don't add `contentful_id` richText.js will return null for references;

**Reply by adstr123** [reply]
This was a big one for me. I had no idea I needed `__typename` in my query from reading the migration guides from v2 or documentation. Is this something that can be done automatically as @axe312ger has mentioned (I don't see an issue logged btw) or is this just something that needs attention brought to it in the documentation?

**Reply by axe312ger** [reply]
@adstr123 all documentation and migration should contain __typename for a while. Where did you look at?

**Comment by gregoralbrecht** [comment]
I'm trying to upgrade to v4 of `gatsby-source-contentful` but can't get it to work. I'm querying the rich text like this (unchanged except for the `raw` instead of `json`):

```graphql
blocks {
  ... on ContentfulBlockRichText {
    body {
      raw
    }
  }
}
```

Which worked fine before the upgrade. If I pass `raw` down to the new `renderRichText(raw, {})`, I get an error:

```
SyntaxError: Unexpected token u in JSON at position 0
    at JSON.parse (<anonymous>)
    at renderRichText (rich-text.js:13)
    at RichText (index.jsx:101)
    at renderWithHooks (react-dom.development.js:14804)
```

Running `JSON.parse(raw)` seems to work perfectly fine, so I've run out of ideas...

**Reply by mikaelkristiansson** [reply]
when passing it to `renderRichText` make sure you pass in the parent and **not** `raw`, so:
```js
renderRichText(body);
```
because it will extract raw from the object you are passing into renderRichText. As you can see here: https://github.com/gatsbyjs/gatsby/blob/master/packages/gatsby-source-contentful/src/rich-text.js#L4

**Reply by gregoralbrecht** [reply]
Well... that was a quick fix, thank you! My bad, I never spotted that difference in the code example, since I thought `renderRichText` was identical to `documentToReactComponents`. Makes sense, though, since it needs `references` as well.

**Reply by axe312ger** [reply]
The readme does not really document that function expect here:

https://github.com/gatsbyjs/gatsby/tree/master/packages/gatsby-source-contentful#rendering

Should I add a section that explains that function and parameters?

**Reply by gregoralbrecht** [reply]
@axe312ger I think having more examples of this might be very helpful! Here are a few things I stumbled upon:
1. The example you linked to has a typo: `richTextField` is not defined, right?
2. It seems like we **_have to_** query `__typename`, too? See the other comment above. Even if this is resolved later, it should probably be in the docs.
3. `pageContext` seems irrelevant?
4. We query `allContentfulBlogPost`, yet receive `data.contentfulBlogPost`? Is that correct?

**Reply by axe312ger** [reply]
> @axe312ger I think having more examples of this might be very helpful! Here are a few things I stumbled upon:
> 
> 1. The example you linked to has a typo: `richTextField` is not defined, right?

I don't know. Thats just a field name. It's up to you how you name your fields in Contentful

> 2. It seems like we **_have to_** query `__typename`, too? See the other comment above. Even if this is resolved later, it should probably be in the docs.

I added it to the readme

> 3. `pageContext` seems irrelevant?

Remove it

> 4. We query `allContentfulBlogPost`, yet receive `data.contentfulBlogPost`? Is that correct?

The query and the code are just examples: I adjusted them to be aligned with each other.


PR for changed readme: https://github.com/gatsbyjs/gatsby/pull/29029

**Comment by gregoralbrecht** [comment]
Does anyone have recommendations on how to handle content types that *might* (or might not) be included in Rich Text? Let's imagine a simple content type `BlogPost` of type Rich Text. It might contain other content types (e.g., `BlogPostImage`, `BlogPostCodeSnippet`, ...) that aren't necessarily included, but might be at some point. Since all content is coming from Contentful, this might change at any time.

**Reply by TomPridham** [reply]
you should be able to do this. the `... on` will only resolve if the fields are present
```
        references {
          ... on ContentfulTable {
            ...ContentfulTableFields
          }
          ... on ContentfulAsset {
            ...ContentfulAssetFields
          }
          ... on ContentfulEmbeddedContent {
            ...ContentfulEmbeddedContentFields
          }
        }
```

**Reply by axe312ger** [reply]
As soon this PR is solved, this issue should be gone: https://github.com/gatsbyjs/gatsby/pull/12816

**Reply by ricardoaguiar** [reply]
I can't get the embed asset to work. I've looked at @Khaledgarbaya examples, and the refactor thread, still can't figure it out what I'm missing. Any help is appreciated. 

```javascript

contentfulPortfolioItem(slug: { eq: $slug }) {
      id
      title
      description: projectDescription {
        raw
        references {
          contentful_id
          file {
            url
            contentType
          }
        }
      }
    }

```

```javascript

  const options = {
    renderNode: {
      [BLOCKS.HEADING_1]: children => <h1>{children}</h1>,
      [BLOCKS.EMBEDDED_ASSET]: node => <Img src={node.description.references.file.url} />
    }
  }

```

**Reply by axe312ger** [reply]
your query is invalid, please adjust your query structure:

```graphql
contentfulPortfolioItem(slug: { eq: $slug }) {
  id
  title
  description: projectDescription {
    raw
    references {
      ... on ContentfulAsset {
        contentful_id
        file {
          url
          contentType
        }
      }
    }
  }
}
```

**Reply by abdojulari** [reply]
It is really sucks. The documentation is a flop. Why is it a big to display an image from contentful?

**Reply by axe312ger** [reply]
@abdojulari Sorry for you disliking the documentation.

Any suggestions? 

What do you mean by `Why is it a big to display an image from contentful?`?

**Reply by EdPike365** [reply]
I just wrote a Medium post on how to fix the "optional embeds in rich text" issue. Here: https://edpike365.medium.com/gatsby-contentful-rich-text-migrate-to-gatsby-source-contentful-version-4-in-early-2021-321904587470

**Reply by aericson** [reply]
> I just wrote a Medium post on how to fix the "optional embeds in rich text" issue. Here: https://edpike365.medium.com/gatsby-contentful-rich-text-migrate-to-gatsby-source-contentful-version-4-in-early-2021-321904587470

@EdPike365 Just wanted to say thanks for the article, was very helpful :)

**Comment by RichardH9L** [comment]
`contentful_id` is sometimes of type `String` and sometimes of type `String!` Is this a bug? If not, what can I do to prevent this from happening?

This is a problem when required to use `contentful_id` on creating references

```
body {
        raw
        references {
          __typename
          ... on ContentfulArticle {
            contentful_id
            slug
          }
          ... on ContentfulVideo {
            contentful_id
            slug
          }
          ...
```

Error:
```
Fields "contentful_id" conflict because they return conflicting types String and String!. Use different aliases on the fields to fetch both if this was intentional.

GraphQL request:56:13
55 |           ... on ContentfulArticle {
56 |             contentful_id
   |             ^
57 |             id

GraphQL request:125:13
124 |           ... on ContentfulVideo {
125 |             contentful_id
    |             ^
126 |             url
```

I worked around it by creating an alias for one of the fields. 

I'm using gatsby 2.28.0 and gatsby-source-contentful@4.2.0, also tried 4.3.0-next.1

**Reply by axe312ger** [reply]
Sounds like a bug, I'll have a look!

**Reply by axe312ger** [reply]
Does this issue still exist for you? I was not able to replicate the issue

**Reply by pfarnach** [reply]
I also ran into this issue -- my schema looks like 
```gql
article {
  raw
  references {
    ... on ContentfulAsset {
      __typename
      contentful_id
      title
      fluid(maxWidth: 1200) {
        src
      }
    }
    ... on ContentfulMyCustomEmbedEntry {
      __typename
      contentful_id
      title
    }
  }
}
```

I also got around it by aliasing the first `contentful_id` and then assigning it to `contentful_id` later in code. 

I'm on my second day of using Gatsby, and between this issue and the issue above that I also ran into (https://github.com/gatsbyjs/gatsby/discussions/28098#discussioncomment-133574), it feels like I'm fighting Gatsby at every turn just to render rich text. What am I missing?

**Reply by axe312ger** [reply]
@pfarnach your query looks fine, what do you mean by ` aliasing the first contentful_id and then assigning it to contentful_id later in code.`?

**Reply by pfarnach** [reply]
> @pfarnach your query looks fine, what do you mean by ` aliasing the first contentful_id and then assigning it to contentful_id later in code.`?

Since `renderRichText` needs the `contentful_id` property to properly resolve references in the raw text (but I can't query for both `contentful_id`s at the same time), the relevant part of my query ends up looking like:
```gql
article {
  raw
  references {
    ... on ContentfulAsset {
      __typename
      contentful_id_alias: contentful_id
      title
      fluid(maxWidth: 1200) {
        src
      }
    }
    ... on ContentfulMyCustomWidget {
      __typename
      contentful_id
      title
    }
  }
}
```
and the code that adds in the `contentful_id` property looks like:
```ts
retVal.references = references?.map(ref => {
  if (ref.contentful_id_alias) {
    ref.contentful_id = ref.contentful_id_alias;
  }
  return ref;
});
```
Pretty gross -- let me know if I'm missing something obvious or if there's a better workaround.

**Reply by axe312ger** [reply]
Querying both without should absolutely work. Looks like this is a bug

**Reply by axe312ger** [reply]
Want to create a ticket out of it?

**Reply by pfarnach** [reply]
I spent some time trying to reproduce this in a smaller project but was unable to. The issue may be related to different plugins not playing together nicely. @RichardH9L can you give us some insight into your particular setup (e.g. what plugins you were using and the general project setup)?

**Comment by abdojulari** [comment]
Thanks for reaching out to me. I am sorry with the hash comment though. I
must start with commendations, you have put up a good documentation.
Nevertheless, it is necessary to update the documents when things change.
For instance, it used to be body { JSON } but not split into body { raw
references { } }. There should be a documentation to that effect that is
straightforward to approach for any developer regardless our levels. I am a
beginner, life was miserable until I found a solution through someone on
YouTube.
Thank you for reaching out and I am very sorry.

Thank you for your time.

Abdul

**Comment by mareksuscak** [comment]
To be quite frank with you all, the new `references` field further extends an already existing and pretty serious flaw of this plugin which is [described in its documentation](https://www.gatsbyjs.com/plugins/gatsby-source-contentful/#notes-on-contentful-content-models) and summarized in the following quote:

> **At the moment, fields that do not have at least one populated instance will not be created in the GraphQL schema.**

Why? I'm glad you asked! Read on.

Before v4 came out, we had to fight this plugin by creating at least one entry for each content type that would populate all optional fields to ensure that the GQL schema generated by Gatsby at build time is stable and doesn't change depending on the content. However, the references inside rich text fields weren't affected by this issue because they were not individually queryable as it was all taken care of by the plugin itself.

After v4, this issue now affects rich text field references as well which means that we not only need to maintain redundant entries with all of the optional fields filled out, now we also need to make sure that all rich text fields that support different kinds of embedded entries, entry hyperlinks or asset hyperlinks need at least one entry that lists out at least one instance of each referenced type which is really insane if you ask me. Consider the following numbers:

1. We use 53 rich text fields across 43 different content types
2. Each rich text field allows one or more of the following reference types: 
  a. entry hyperlinks
  b. asset hyperlinks
  c. embedded entries
3. We use different validation rules for the above reference types allowing certain types of entries to be embedded

As you can imagine, it'd be a TON of work to ensure GQL schema stability with such a large and dynamic content model.

Don't get me wrong, it would be a huge improvement over the previous solution if that wasn't the case, but given the aforementioned drawback, it's become unfeasible for us to migrate to newer versions due to all the extra work. Are there any plans to improve the situation and generate the schema from the validation rules as that's the only way to ensure stable schema? We can't be the only ones suffering from this issue or are we outliers? Is there anyone else who shares my point of view? Oddly enough, GraphQL API provided by Contentful itself is always generated correctly based on validation rules and not content. So what is preventing this plugin from doing the same or utilizing the existing GQL schema? 

I'm not familiar enough with the implementation details and limitations and so I can't tell if there's a serious blocker preventing this from being implemented correctly but I'd be happy to take a stab at it if someone who's more familiar provides some context and pointers.

**Reply by gregoralbrecht** [reply]
Thanks for writing this! To be honest, I'm relieved to read this—I started to think I was doing something fundamentally wrong because I haven't seen more people stumble over this.

I'm sure there are excellent reasons for this behavior, and I'm also not familiar with the implementation details. So I can't really tell if this is something that can be tackled by this plugin or not. However, this behavior results in a huge work overhead for us, too. 

Right now, I'm trying to work around this with a pretty ugly hack involving [`gatsby-plugin-schema-snapshot`](https://github.com/gatsbyjs/gatsby/tree/master/packages/gatsby-plugin-schema-snapshot):

0. I'm building "blocks" in the Gatsby project that map directly to specific content types in Contentful. The goal is the let content editors add/remove/update these without needing a developer.
1. Whenever I need to add such a block, I create a new content type and then create a dummy content entry in Contentful with all fields filled out (without publishing it). Then, I run the Gatsby project with the above-mentioned plugin and the [content preview API](https://www.contentful.com/developers/docs/references/content-preview-api/). 
2. I then need to find the parts in the generated schema snapshot that represents the new content type and commit these with the snapshot.
3. Then, I remove the dummy content entry from Contentful.
4. Next time I run the plugin, I need to make sure not to remove the parts that represent the now-removed content type.

This is so hacky and fragile that I hate admitting this. But if I don't do this, I'm running a huge chance of our CI pipeline breaking or our colleagues not being able to work just because a content editor removed the single entry that was needed to add this to the schema.

**Reply by axe312ger** [reply]
@gregoralbrecht I know its complicated, but the way you do it, is a working solution for now.

Regarding schema introspection of the plugin:

* Yes, the automatically generated schema of the plugin is problematic as it depends on the actual content, not the content model from Contentful
* Yes, Contentfuls GraphQL API does supply a schema, but gatsby-source-contentful does NOT use the Contentful GraphQL API. The plugin is way older as the API
* We are preparing a rewrite, moving the plugin to Contentfuls GraphQL API, which will ultimately fix a lot of issues, especially because we can adopt the schema supplied by Contentful.
* Stay tuned, can't give you any ETA yet

**Reply by mareksuscak** [reply]
Thanks for your response @axe312ger. I really appreciate it. Is there anything we as a community can do to contribute to the development of the upgraded version? Is there a single place where we can start having meaningful conversations about the tradeoffs and possible migration path?

**Reply by axe312ger** [reply]
Not yet, I'll open a new issue as soon we start working on it so the community can collaborate!

**Reply by tomlschmidt** [reply]
Do we have any update on this @axe312ger? It still is a substantial issue in my opinion.

**Reply by axe312ger** [reply]
@tomlschmidt see #31385

I agree with you but it is far from easy to fix. I expect to have public updates soon!

**Reply by healthierkamm** [reply]
> I know its complicated, but the way you do it, is a working solution for now.

The GraphQL I'm using, which queries for linked entries, breaks when there isn't any content that actually links to an entity link. Is the recommended solution still to populate dummy content?

**Reply by axe312ger** [reply]
Yes, till #31385 is released 🙈

**Reply by healthierkamm** [reply]
hey, good to know; thanks for the reply

**Comment by awkale** [comment]
I've upgraded and switched my queries using `raw` but keep getting this
![image](https://user-images.githubusercontent.com/744487/111009036-810acd80-8360-11eb-9fc1-e49393854b3c.png)

Also, was using `import { documentToReactComponents } from '@contentful/rich-text-react-renderer'` before the upgrade and now use `import { renderRichText } from 'gatsby-source-contentful/rich-text'` 

here's my query and how I'm using it
```
contentfulMainDescription(page: { slug: { eq: "product-updates" } }) {
      name
      text {
        raw
      }
    }

{renderRichText(data.contentfulMainDescription.text)}
```

Has anyone seen this error?

**Reply by axe312ger** [reply]
@awkale what is the value of `data.contentfulMainDescription.text`?

You might want to check if it actually has a value before trying to render it
